docs/contributing/code-reviewing.md

```diff
@@ -1,231 +1,364 @@
-  conflicts) if there are still user experience issues under
-  discussion for the feature itself.
-
-- _Completeness._ When reviewing a refactor, verify that the changes are
+- _Completeness of refactoring._ When reviewing a refactor, verify that the changes are
   complete. Usually, one can check that efficiently using `git grep`,
   and it's worth it, as we very frequently find issues by doing so.
 
 - _Documentation updates._ If this changes how something works, does it
   update the documentation in a corresponding way? If it's a new
   feature, is it documented, and documented in the right place?
 
-- _Good comments._ It's often worth thinking about whether explanation
-  in a commit message or pull request discussion should be included in
-  a comment, `/docs`, or other documentation. But it's better yet if
-  verbose explanation isn't needed. We prefer writing code that is
-  readable without explanation over a heavily commented codebase using
-  lots of clever tricks.
+- _mypy annotations in Python code._ New functions should be annotated using
+  [mypy] and existing annotations should be updated. Use of `Any`, `ignore`, and
+  unparameterized containers should be limited to cases where a more precise
+  type cannot be specified.
+
+### Automated testing
+
+- The tests should **validate that the feature works correctly**, and
+  specifically test for common error conditions, bad user input, and potential
+  bugs that are likely for the type of change being made. Tests that exclude
+  whole classes of potential bugs are preferred when possible (e.g., the common
+  test suite `test_markdown.py` between the Zulip server's [frontend and backend
+  Markdown processors](../subsystems/markdown.md), or the `GetEventsTest` test
+  for buggy race condition handling). See the [test writing][test-writing]
+  documentation to learn more.
+
+- We are trying to maintain ~100% test coverage on the backend, so backend
+  changes should have negative tests for the various error conditions. See
+  [backend testing documentation][backend-testing] for details.
+
+- If the feature involves frontend changes, there should be frontend tests. See
+  [frontend testing documentation][frontend-testing] for details.
+
+### Manual testing
+
+If the PR makes any frontend changes, you should make sure to play with the part
+of the app being changed to validate that things look and work as expected.
+While not all of the situations below will apply, here are some ideas for things
+that should be tested if they are applicable. Use the [development
+environment][development-environment] to test any webapp changes.
+
+This might seem like a long process, but you can go through it quite quickly
+once you get the hang of it. Trust us, it will save time and review round-trips
+down the line!
+
+**Visual appearance:**
+
+- Open up the parts of the UI that were changed, and make sure they look as
+  you were expecting.
```

Comment:
Should we mention using your browser inspector to check a range of window sizes?

Reply:
I guess this is below, OK.